File: tests/unit/relays/orchestration/hydration-priority.test.ts
import { describe, expect, it } from 'vitest';
import {
  planHydrationJobs,
  type HydrationJob,
} from '../../../../src/lib/relays/orchestration/hydration-priority';

describe('hydration priority', () => {
  it('runs visible and near-visible jobs before background work', () => {
    const plan = planHydrationJobs({
      generation: 9,
      jobs: [
        job('background', 'background', 9),
        job('near', 'near-visible', 9),
        job('visible', 'visible', 9),
      ],
    });

    expect(plan.runnable.map((item) => item.key)).toEqual([
      'visible',
      'near',
      'background',
    ]);
  });

  it('dedupes by semantic key and cancels stale jobs', () => {
    const plan = planHydrationJobs({
      generation: 4,
      jobs: [
        job('profile:alice', 'background'),
        job('profile:alice', 'visible'),
        job('profile:bob', 'hidden-paused'),
        { ...job('profile:carol', 'visible'), generation: 3 },
      ],
    });

    expect(plan.runnable).toMatchObject([
      { key: 'profile:alice', priority: 'visible' },
    ]);
    expect(plan.cancelled).toEqual(['profile:carol']);
  });
});

function job(
  key: string,
  priority: HydrationJob['priority'],
  generation = 4,
): HydrationJob {
  return { key, priority, generation };
}

File: crates/lkjstr-app/tests/hydration_priority_test.rs
use lkjstr_app::{HydrationJob, HydrationPriority, plan_hydration_jobs};

#[test]
fn visible_jobs_run_before_near_and_background() {
    let (jobs, cancelled) = plan_hydration_jobs(
        &[
            job("background", HydrationPriority::Background, 7),
            job("near", HydrationPriority::NearVisible, 7),
            job("visible", HydrationPriority::Visible, 7),
        ],
        7,
    );

    assert_eq!(keys(&jobs), vec!["visible", "near", "background"]);
    assert!(cancelled.is_empty());
}

#[test]
fn hidden_stale_and_duplicate_jobs_do_not_run() {
    let (jobs, cancelled) = plan_hydration_jobs(
        &[
            job("same", HydrationPriority::NearVisible, 3),
            job("same", HydrationPriority::Visible, 3),
            job("hidden", HydrationPriority::HiddenPaused, 3),
            job("stale", HydrationPriority::Visible, 2),
        ],
        3,
    );

    assert_eq!(keys(&jobs), vec!["same"]);
    assert_eq!(cancelled, vec!["stale"]);
}

fn job(key: &str, priority: HydrationPriority, generation: u64) -> HydrationJob {
    HydrationJob {
        key: key.to_owned(),
        priority,
        generation,
    }
}

fn keys(jobs: &[HydrationJob]) -> Vec<&str> {
    jobs.iter().map(|job| job.key.as_str()).collect()
}
